src/mlpack/core/cereal/main.cc

```diff
@@ -0,0 +1,35 @@
+#include "pointer_wrapper.hpp"
+#include <fstream>
+
+void ref_to_pointer(double*& p)
+{
+  if(p == nullptr) {
+    p = new double;
+  }
+  std::cout << "address of  pointer" << p << std::endl;
+  
+}
+
+int main(int argc, char* argv[])
+{
+
+  std::ifstream is("data.xml");
+  cereal::XMLInputArchive archive(is);
+  
+  // std::ofstream is("data.xml");
+  // cereal::XMLOutputArchive archive(is);
+  
+  double data = 40;
+  double* d = nullptr;
+
+  //ref_to_pointer(d); 
+
+  archive(cereal::make_pointer(d));
```

Comment:
In order to deserialize properly, I had to use this code:

```
  cereal::pointer_wrapper<double> d = cereal::make_pointer(z);
  archive(d);
  z = d.localPointer;

  if ( z == NULL ) {
    std::cout << "z is still NULL" << std::endl;
  } else {

    std::cout << "z value is : " << *z << std::endl;
  }
```

When I do that (and when data.xml properly contains the value `40`), the output I see when I run the program is:

```
address of local ptr0
address of  pointer0
The value of localPointer is : 40
The addresss of localPointer is : 0x7ffc32f70778
z value is : 40
```

I think @robertohueso is right that this can be solved with double pointers (or a pointer reference): https://github.com/mlpack/mlpack/pull/2415#discussion_r441902498

Reply:
@rcurtin I thought of this solution to recover directly the address of the `localPointer` using a function, but in this case, I will have to revisit all the cases where the pointer is used and recover it manually